### PR TITLE
DM-28859: Make the DBB endpoint buffer manager persist file sizes in the database

### PR DESCRIPTION
**Pull request by mxk62**
Besides changes directly related to the feature request, I made some additional modifications. Most of them are cosmetic ones to make pylint complain a little bit less about the code. For example, I added few missing module docstrings and removed unnecessary `pass` statements.

However, I also replaced each component's `--validate` option with the corresponding, standalone command. The option was effectively turning each component into a validation tool as setting it on was making each component to terminate its execution *regardless* of the outcome of the validation. So I hope that capturing this functionality in a separate subcommand will provide a better "separation of concerns".

### CODE REVIEW

**Comment by mxk62** [comment]
Besides changes directly related to the feature request, I made some additional modifications. Most of them are cosmetic ones to make `pylint` complain a little bit less about the code. For example, I added few missing module docstrings and removed unnecessary `pass` statements.

However, I also replaced each component's `--validate` option with the corresponding, standalone command. The `--validate` option was effectively turning each component into a validation tool as setting it on was making each component to terminate its execution *regardless* of the outcome of the validation. So I hope that capturing this functionality in a separate subcommand will provide a better "separation of concerns".

File: python/lsst/dbb/buffmngrs/endpoint/backfill/backfill.py [hunk 1]
```diff
@@ -144,9 +143,9 @@ def run(self):
                     counts["tracked"] += 1
                     continue
 
-                logger.debug("calculating checksum")
+                logger.debug("getting file attributes")
                 try:
-                    checksum = get_checksum(abspath)
+                    checksum, status = get_file_attributes(abspath)
```
Comment (MichelleGower) [inline python/lsst/dbb/buffmngrs/endpoint/backfill/backfill.py]: To make this more future proof, really should return a container of all file attributes and pass that container (perhaps modified) to a function to save the file attributes.  Having each subcommand know exactly how many attributes there are   lessens the helpfulness of the get_file_attributes function.

File: python/lsst/dbb/buffmngrs/endpoint/cli.py [hunk 2]
```diff
@@ -18,8 +18,10 @@
 #
 # You should have received a copy of the GNU General Public License
 # along with this program.  If not, see <https://www.gnu.org/licenses/>.
-import click
+"""Command line interface for the DBB endpoint buffer manager.
+"""
 import logging
+import click
 from .backfill.commands import backfill
```
Comment (MichelleGower) [inline python/lsst/dbb/buffmngrs/endpoint/cli.py:27]: PEP8 recommends: "You should put a blank line between each group of imports."

File: python/lsst/dbb/buffmngrs/endpoint/backfill/commands.py [hunk 3]
```diff
@@ -82,3 +75,14 @@ def start(filename, dump, validate):
...
+
+@backfill.command()
+@click.argument("filename", type=click.Path(exists=True))
+def validate(filename):
+    """Validate configuration in the FILENAME.
+    """
+    with open(filename) as f:
+        configuration = yaml.safe_load(f)
+    schema = yaml.safe_load(BACKFILL)
+    validate_config(configuration, schema)
```
Comment (MichelleGower) [inline python/lsst/dbb/buffmngrs/endpoint/backfill/commands.py]: Lines 85-87 are repeated with every validate_config call.  Is there a way to move them into the validate_config function?

File: python/lsst/dbb/buffmngrs/endpoint/ingester/ingester.py [hunk 4]
```diff
@@ -208,17 +213,16 @@ def run(self):
...
-                        status = Status.INVALID
-                if status == Status.INVALID:
-                    logger.warning("'%s':  %s", path, message)
+                logger.debug("%s: size is %i bytes", path, size)
+                if not os.path.isfile(os.path.join(self.storage, path)):
+                    message = f"no such file in the storage area"
+                    status = Status.INVALID
+                if message is not None:
+                    logger.debug("%s: message: '%s'", path, message)
+                    logger.debug("%s: status: '%s'", path, status.value)
```
Comment (MichelleGower) [inline python/lsst/dbb/buffmngrs/endpoint/ingester/ingester.py]: Consider whether better to combine these two pieces of info into a single debugging message for easier searching.

File: python/lsst/dbb/buffmngrs/endpoint/ingester/plugins.py [hunk 5]
```diff
@@ -175,7 +177,13 @@ def __init__(self, config):
         ingest_class = doImport(self.config["ingest_task"])
         ingest_config = ingest_class.ConfigClass()
         ingest_config.transfer = self.config["transfer"]
-        ingest_config.failFast = self.config["failFast"]
+
+        # TODO: A temporary hack to make the plugin usable with LSST stack
+        #  prior to w_2021_05. Remove the if statement once a later version is
+        #  in use.
```
Comment (MichelleGower) [inline python/lsst/dbb/buffmngrs/endpoint/ingester/plugins.py]: Better wording (i.e., the very first failFast doesn't guarantee we wouldn't roll back):  once the pre-failFast stack versions are old enough to never be used again.

File: python/lsst/dbb/buffmngrs/endpoint/status.py [hunk 6]
```diff
@@ -18,6 +18,8 @@
 #
 # You should have received a copy of the GNU General Public License
 # along with this program.  If not, see <https://www.gnu.org/licenses/>.
+"""A list of available ingest statues.
```
Comment (MichelleGower) [inline python/lsst/dbb/buffmngrs/endpoint/status.py]: typo:  statues -> statuses

File: etc/backfill.yaml [hunk 7]
```diff
@@ -13,8 +13,7 @@ backfill:
   storage: <storage area>
   sources:
     - <file source>
-  search:
-    exclude_list: null
+  exclude_list: null
```
Comment (MichelleGower) [inline etc/backfill.yaml]: Now only finder has search subsection with exclude_list in this subsection.  Consider whether worth having them at different level in finder than the other components.

Comment (MichelleGower) [reply etc/backfill.yaml]: Also the search section still exists in the config doc for backfill.

File: python/lsst/dbb/buffmngrs/endpoint/finder/finder.py [hunk 8]
```diff
@@ -167,6 +168,7 @@ def run(self):
                     relpath=os.path.relpath(dirname, start=self.storage),
                     filename=basename,
                     checksum=checksum,
+                    size_bytes=status.st_size,
```
Comment (MichelleGower) [inline python/lsst/dbb/buffmngrs/endpoint/finder/finder.py]: Same comment as in backfill about better to have a function for the db operation to allow easier additions of more columns.

**Review by MichelleGower — APPROVED** [review]
A couple minor things and perhaps adding a db function to more future proof code.  Merge approved.

File: etc/backfill.yaml (hunk 7 above)
Comment (mxk62) [reply etc/backfill.yaml]: I'll bring back the `search` section to the backfill config.

File: python/lsst/dbb/buffmngrs/endpoint/backfill/backfill.py (hunk 1 above)
Comment (mxk62) [reply python/lsst/dbb/buffmngrs/endpoint/backfill/backfill.py]: I don't think I quite understand what is your point here. It already returns a container, a tuple. You could pass it around without worrying about number of elements by writing, say,
```
attrs = get_file_attributes(abspath)
```
(can't modify it though without converting it to a list first). It just happens that I need to access its elements shortly after so I'm unpacking it "in-place". Do you want me to change the type of the container to something else?

Comment (MichelleGower) [reply python/lsst/dbb/buffmngrs/endpoint/backfill/backfill.py]: Say next week we want to save another value per file (newattr1).  With the current code, everywhere get_file_attributes is called (which is more than this one line) the code would have to be changed to include newattr1 in the return tuple.   

If one has code similar in concept to:
```
mydict = get_file_attributes(abspath)
<code here that actually must use/modify certain mydict values>
add_file(self.session, mydict)
```
it doesn't matter here if mydict changes. The only changes in this case would go in the shared get_file_attributes and add_file functions.

**Pull request merged**